mixin/smarty-v2@1/mixin.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+/**
+ * Auto-register "templates/" folder.
+ *
+ * @mixinName smarty-v2
+ * @mixinVersion 1.0.0
+ * @since 5.58
+ *
+ * @param CRM_Extension_MixInfo $mixInfo
+ *   On newer deployments, this will be an instance of MixInfo. On older deployments, Civix may polyfill with a work-a-like.
+ * @param \CRM_Extension_BootCache $bootCache
+ *   On newer deployments, this will be an instance of MixInfo. On older deployments, Civix may polyfill with a work-a-like.
+ */
+return function ($mixInfo, $bootCache) {
+  $dir = $mixInfo->getPath('templates');
+  if (!file_exists($dir)) {
+    return;
+  }
+
+  $register = function() use ($dir) {
+    // This implementation is useful for older versions of CiviCRM. It can be replaced/updated going forward.
+    $smarty = CRM_Core_Smarty::singleton();
+    if (!is_array($smarty->template_dir)) {
+      $this->template_dir = [$smarty->template_dir];
+    }
+    if (!in_array($dir, $smarty->template_dir)) {
+      array_unshift($smarty->template_dir, $dir);
+    }
+  };
+
+  if ($mixInfo->isActive()) {
+    \Civi::dispatcher()->addListener('hook_civicrm_config', $register);
+  }
+  elseif (CRM_Extension_System::singleton()->getManager()->extensionIsBeingInstalledOrEnabled($mixInfo->longName)) {
+    $register();
```

Comment:
When I took the first draft and tried to add a test, I found that it didn't work well for the install/disable processes - because the older `hook_config` path was involved during setup. There was a similar issue when @colemanw and I were trying to figure a change for the `include_path`. 

This funny function (`extensionIsBeingInstalledOrEnabled()`) goes back to 5.25 (20429eb9408fc3696d8b2c775bc6b7020a56eaf8), and... in the E2E test... it does seem to work...

Reply:
hmm - OK so reading that it's pretty clear it's saying that if the extension is installed or being enabled it should register something.

I'm not too sure what is happening in the first part of the if there - ie what does it mean for mixInfo to be active & I don't really understand how the listeners would use the inline function / why it is an inline function not a function with a comment block - but I'm just reading that one file without having gone off reading the other files to understand what the MixInfo because you commented on one part of it. 

Overall I don't have to follow this bit of code I don't think - it's enough that it makes sense to you.

Reply:
> I'm not too sure what is happening in the first part of the if there - ie what does it mean for mixInfo to be active & I don't really understand how the listeners would use the inline function 

Added some more inline comments. Don't know if it's enough for every reader, but hopefully it's a bit more explanatory.

> why it is an inline function not a function with a comment block 

I'd say that inline functions and inline classes are typical in the "mixin" files. Globally named functions+classes create a point of contention. This technique makes it safer to mix/scan across many different versions.

Think of the code-style as "less Java" and "more NodeJS".

> Overall I don't have to follow this bit of code I don't think - it's enough that it makes sense to you.

They're good questions above, but this is also valid. 🙃